core/benches/bench_lazy.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use criterion::{
+    black_box,
+    criterion_group,
+    criterion_main,
+    BenchmarkId,
+    Criterion,
+};
+use ink_core::{
+    env,
+    storage2::{
+        lazy::Lazy,
+        traits::{
+            KeyPtr,
+            SpreadLayout,
+        },
+    },
+};
+use ink_primitives::Key;
+
+criterion_group!(populated_cache, bench_set_populated_cache);
+criterion_group!(empty_cache, bench_set_empty_cache,);
+criterion_main!(populated_cache, empty_cache);
+
+mod populated_cache {
+    use super::*;
+    use core::ops::DerefMut;
+
+    pub fn set() {
+        let mut lazy = <Lazy<i32>>::new(1);
+        let lazy_mut = black_box(&mut lazy);
+        black_box(Lazy::set(lazy_mut, 17));
+    }
+
+    pub fn deref_mut() {
+        let mut lazy = <Lazy<i32>>::new(1);
+        let i32_mut = black_box(lazy.deref_mut());
+        black_box(*i32_mut = 17);
+    }
+}
+
+fn bench_set_populated_cache(c: &mut Criterion) {
+    let mut group = c.benchmark_group("Compare: `set` and `deref_mut` (populated cache)");
+    group.bench_function(BenchmarkId::new("set", 0), |b| {
+        b.iter(|| populated_cache::set())
+    });
+    group.bench_function(BenchmarkId::new("deref_mut", 0), |b| {
+        b.iter(|| populated_cache::deref_mut())
+    });
+    group.finish();
+}
+
+/// Pushes a value to contract storage and creates a `Lazy` pointing to it.
+fn push_storage_lazy(value: i32) -> Lazy<i32> {
+    let root_key = Key::from([0x00; 32]);
+    SpreadLayout::push_spread(&Lazy::new(value), &mut KeyPtr::from(root_key));
+    <Lazy<i32>>::lazy(root_key)
+}
+
+mod empty_cache {
+    use super::*;
+
+    pub fn set() {
+        let mut lazy = push_storage_lazy(1);
+        black_box(Lazy::set(&mut lazy, 13));
+    }
+
+    pub fn deref_mut() {
+        let mut lazy = push_storage_lazy(1);
+        black_box(*lazy = 13);
```

Comment:
```suggestion
        black_box(*lazy) = 13;
```

Reply:
Please try this and see if it changes something in the benchmarks.

Reply:
We want to keep the scope of our `black_box` as big as needed but as small as possible.

Reply:
```
error[E0070]: invalid left-hand side of assignment
  --> core/benches/bench_lazy.rs:93:26
   |
93 |         black_box(*lazy) = 13;
   |         ---------------- ^
   |         |
   |         cannot assign to this expression
```
I had tried to get variations of this to work as well. So seems like this is the smallest scope to which we can get.